rcl/include/rcl/client.h

```diff
@@ -0,0 +1,300 @@
+
+/// Send a ROS request using a client.
+/* It is the job of the caller to ensure that the type of the ros_request
+ * parameter and the type associate with the client (via the type support)
+ * match.
+ * Passing a different type to send_request produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * send_request is an non-blocking call.
+ *
+ * The ROS request message given by the ros_request void pointer is always owned by the
+ * calling code, but should remain constant during send_request.
+ *
+ * This function is thread safe so long as access to both the client and the
+ * ros_request is synchronized.
+ * That means that calling rcl_send_request from multiple threads is allowed, but
+ * calling rcl_send_request at the same time as non-thread safe client functions
+ * is not, e.g. calling rcl_send_request and rcl_client_fini concurrently
+ * is not allowed.
+ * Before calling rcl_send_request the message can change and after calling
+ * rcl_send_request the message can change, but it cannot be changed during the
+ * send_request call.
+ * The same ros_request, however, can be passed to multiple calls of
+ * rcl_send_request simultaneously, even if the clients differ.
+ * The ros_request is unmodified by rcl_send_request.
+ *
+ * \param[in] client handle to the client which will make the response
+ * \param[in] ros_request type-erased pointer to the ROS request message
+ * \return RCL_RET_OK if the request was sent successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_CLIENT_INVALID if the client is invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_send_request(const rcl_client_t * client, const void * ros_request);
+
+
+// Take a ROS response using a client
+/* It is the job of the caller to ensure that the type of the ros_response
+ * parameter and the type associate with the client (via the type support)
+ * match.
+ * Passing a different type to take_response produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ * The request_header is an rmw struct for meta-information about the request sent
+ * (e.g. the sequence number).
+ * ros_response should point to an already allocated ROS response message struct of the
+ * correct type, into which the response from the service will be copied.
+ *
+ * \param[in] client handle to the client which will take the response
+ * \param[inout] request_header pointer to the request header
+ * \param[inout] ros_response type-erased pointer to the ROS response message
+ * \return RCL_RET_OK if the response was taken successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_CLIENT_INVALID if the client is invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
```

Comment:
`RCL_RET_CLIENT_TAKE_FAILED` doesn't seem to be documented here. Also did we ever decide if that was better (preferable) to passing a `bool * taken` parameter?